crates/accelerate/src/circuit_library/entanglement.rs

```diff
@@ -0,0 +1,197 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use itertools::Itertools;
+use pyo3::{
+    types::{PyAnyMethods, PyInt, PyList, PyListMethods, PyString, PyTuple},
+    Bound, PyAny, PyResult,
+};
+
+use crate::QiskitError;
+
+/// Get all-to-all entanglement. For 4 qubits and block size 2 we have:
+/// [(0, 1), (0, 2), (0, 3), (1, 2), (1, 3), (2, 3)]
+pub fn full(num_qubits: u32, block_size: u32) -> impl Iterator<Item = Vec<u32>> {
```

Comment:
I am trying to understand when a function can return `impl Iterator<Item = Vec<u32>>` vs `Box<dyn Iterator<Item = Vec<u32>>>`. From what I see in the documentation, we can do the former when the concrete iterator type is known at compile time, and we should do the latter when we want to return multiple iterator types. But why does `full` fall into the first category, while `shift_circular_alternating` into the second one?

In any case, would it be better if all the functions returned the same type, be it `Box<dyn Iterator<Item = Vec<u32>>>` or maybe even `Box<dyn DoubleEndedIterator<Item = Vec<u32>>>`?

Reply:
As discussed in a comment above, we could also avoid using `DoubleEndedIterator`, but we would still need the `Box<dyn Iterator>` since the functions return different iterator types (they use `Chain` `Map` `Slice` in different combinations, so would still need to be coerced to the same dynamic type). But yeah, we could make all functions return `Box<dyn Iterator<Item=Vec<u32>>` -- I kept the signatures as the minimal type we need, but I'm also happy to change 🙂

Reply:
Thanks! I think the current way is fine. Oh, one more question: do the functions like `full`, `linear` need to be public (`pub`)?

Reply:
They currently are since the `get_entangler_map` with the Python logic is defined in `mod.rs` -- we can move that into the `entanglement.rs` file and then they can be private 🙂